browser-sign-in/app/src/main/java/com/okta/android/samples/browser_sign_in/views/LoginFragment.kt

```diff
@@ -0,0 +1,85 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.okta.android.samples.browser_sign_in.views
+
+import android.os.Bundle
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import androidx.fragment.app.Fragment
+import androidx.fragment.app.viewModels
+import androidx.navigation.fragment.findNavController
+import com.okta.android.samples.browser_sign_in.data.LoginState
+import com.okta.android.samples.browser_sign_in.data.LoginViewModel
+import com.okta.android.samples.browser_sign_in.databinding.FragmentLoginBinding
+import timber.log.Timber
+
+class LoginFragment : Fragment() {
+    private val viewModel: LoginViewModel by viewModels()
+    private lateinit var binding: FragmentLoginBinding
+
+    override fun onCreateView(
+        inflater: LayoutInflater, container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View {
+        binding = FragmentLoginBinding.inflate(layoutInflater, container, false)
+        return binding.root
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        binding.loginButton.setOnClickListener {
+            viewModel.login(requireContext())
+        }
+
+        /**
+         * Update the user interface for changes in the sign-in flow.
+         *
+         * Use an observer to react to updates in [LoginState]. Updates are asynchronous and are triggered both by user actions,
+         * such as button clicks, and completing the flow.
+         */
+        viewModel.loginState.observe(viewLifecycleOwner) { state ->
+            when (state) {
+                is LoginState.Loading -> {
+                    with(binding) {
+                        loginLoadingProgressBar.visibility = View.VISIBLE
+                        loginContainer.visibility = View.GONE
+                    }
+                }
+                is LoginState.LoggedOut -> {
+                    with(binding) {
+                        loginLoadingProgressBar.visibility = View.GONE
+                        loginContainer.visibility = View.VISIBLE
+                    }
+                }
+                is LoginState.LoggedIn -> {
+                    val navigationAction =
+                        LoginFragmentDirections.actionLoginFragmentToUserDashboardFragment()
+                    findNavController().navigate(navigationAction)
+                }
+                is LoginState.LoginFailed -> {
+                    with(binding) {
+                        loginLoadingProgressBar.visibility = View.GONE
+                        loginContainer.visibility = View.VISIBLE
+                        statusTextView.text = state.errorMessage
+                    }
+                }
+                is LoginState.LogoutFailed -> Timber.d("Reached an invalid LoginState in LoginFragment: $state")
```

Comment:
If we have an invalid state, we might not be modeling this correctly. It might be worth separating this state from the one used in the dashboard.